Web Inspector: should not include `@globalPrivate` functions in stack traces

dcrousso opened the pull request:
#### ebb980b89e945e8c3997ecb3c760a1d66c0a901b
<pre>
Web Inspector: should not include `@globalPrivate` functions in stack traces
<a href="https://bugs.webkit.org/show_bug.cgi?id=242806">https://bugs.webkit.org/show_bug.cgi?id=242806</a>

Reviewed by Yusuke Suzuki.

`@globalPrivate` are just internal implementation details, and therefore are likely to change (at
least more often than standardized functions) and be unknown to a developer.

* Source/JavaScriptCore/runtime/ImplementationVisibility.h: Added.
Create a new enum that (currently) allows indicating `Public` or `Private`.

* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_generator.py:
(BuiltinsGenerator.generate_embedded_code_data_for_function):
(BuiltinsGenerator.generate_embedded_code_string_section_for_data):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_generate_combined_header.py:
(BuiltinsCombinedHeaderGenerator.generate_forward_declarations):
(BuiltinsCombinedHeaderGenerator.generate_externs_for_object):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_generate_separate_header.py:
(BuiltinsSeparateHeaderGenerator.generate_externs_for_object):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_templates.py:
Mark functions annotated with `@globalPrivate` as being `ImplementationVisibility::Private`.

* Source/JavaScriptCore/builtins/BuiltinUtils.h:
* Source/JavaScriptCore/builtins/BuiltinExecutableCreator.h:
* Source/JavaScriptCore/builtins/BuiltinExecutableCreator.cpp:
(JSC::createBuiltinExecutable):
* Source/JavaScriptCore/builtins/BuiltinExecutables.h:
* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::createDefaultConstructor):
(JSC::BuiltinExecutables::createBuiltinExecutable):
(JSC::BuiltinExecutables::createExecutable):
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.cpp:
(JSC::UnlinkedFunctionExecutable::UnlinkedFunctionExecutable):
Pass along the `ImplementationVisibility`.

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::makeFunction):
Assume `ImplementationVisibility::Public` for now.

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitNewClassFieldInitializerFunction):
Mark as `ImplementationVisibility::Private` as this is not supposed to be visible to developers.

* Source/JavaScriptCore/runtime/CodeCache.cpp:
(JSC::CodeCache::getUnlinkedGlobalFunctionExecutable):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(functionCreateBuiltin):
Mark as `ImplementationVisibility::Public`.

* Source/JavaScriptCore/runtime/ExecutableBase.h:
* Source/JavaScriptCore/runtime/ExecutableBaseInlines.h:
(JSC::ExecutableBase::implementationVisibility const): Added.
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
(JSC::FunctionExecutable::implementationVisibility const): Added.
Add a way to access the `ImplementationVisibility`.

* Source/JavaScriptCore/inspector/ScriptCallStackFactory.cpp:
(Inspector::CreateScriptCallStackFunctor::operator() const):
Ignore anything marked as `ImplementationVisibility::Private`.

* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.h:
* Source/JavaScriptCore/parser/Nodes.h:
* Source/JavaScriptCore/parser/ParserModes.h:
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
* Source/JavaScriptCore/runtime/ScriptExecutable.h:
Drive-by: Decrease the bit width of `LexicalScopeFeatures` to make room for `ImplementationVisibility`.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:

* LayoutTests/inspector/canvas/recording-bitmaprenderer-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-bitmaprenderer-full-expected.txt:
* LayoutTests/inspector/canvas/recording-bitmaprenderer-memoryLimit-expected.txt:
* LayoutTests/inspector/console/message-stack-trace-expected.txt:
* LayoutTests/inspector/debugger/js-stacktrace-expected.txt:

Canonical link: <a href="https://commits.webkit.org/252554@main">https://commits.webkit.org/252554@main</a>
</pre>

Constellation commented on Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:
```diff
@@ -262,6 +264,7 @@ class UnlinkedFunctionExecutable final : public JSCell {
     unsigned m_lineCount : 31;
     unsigned m_hasCapturedVariables : 1;
     unsigned m_unlinkedFunctionNameStart : 31;
+    unsigned m_implementationVisibility : bitWidthOfImplementationVisibility;
```
Can you make a room in this class and put a flag? Right now, this is 96byte, so adding this 1 bit enlarges it by 16bytes. And this is super frequently allocated class.

Let's modify ParserModes.h's LexicalScopeFeatures related codes to make it 3 bits (right now, it is 4 bits), and allocate 1 bit room.

Constellation replied:
Let's move this to next to m_lexicalScopeFeatures to make it packed well.

Constellation replied:
Let's also add

```
#if COMPILER(CLANG) && !ASSERT_ENABLED
static_assert(sizeof(UnlinkedFunctionExecutable) <= 96, "UnlinkedFunctionExecutable needs to be small");
#endif
```

in this file to ensure this critical condition.

Constellation commented on Source/JavaScriptCore/parser/ParserModes.h:
```diff
@@ -314,6 +314,8 @@ const LexicalScopeFeatures StrictModeLexicalFeature               = 1 << 0;
 const LexicalScopeFeatures AllLexicalFeatures = NoLexicalFeatures | StrictModeLexicalFeature;
 static_assert(AllLexicalFeatures <= 0b1111, "LexicalScopeFeatures must be 4bits");
```
Make it to 0b111 and change comment to say 3 bits.

Constellation reviewed (approved):
r=me with comments. And please ensure that UnlinkedFunctionExecutable size is 96 bytes before landing. I think adding proposed static_assert will easily ensure it.

Constellation commented on Source/JavaScriptCore/parser/ParserModes.h:
```diff
@@ -312,7 +312,8 @@ const LexicalScopeFeatures NoLexicalFeatures                           = 0;
 const LexicalScopeFeatures AllLexicalFeatures = NoLexicalFeatures | StrictModeLexicalFeature;
-static_assert(AllLexicalFeatures <= 0b1111, "LexicalScopeFeatures must be 4bits");
+static constexpr unsigned bitWidthOfLexicalScopeFeatures = 3;
+static_assert(AllLexicalFeatures <= bitWidthOfLexicalScopeFeatures, "LexicalScopeFeatures must be 3bits");
```
Ah, I think it should be `(1 << bitWidthXXX) - 1`

dcrousso replied:
oh lol oops

JosephPecoraro commented on Source/JavaScriptCore/runtime/ImplementationVisibility.h:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C) 2019 Apple Inc. All Rights Reserved.
```
Me thinks you're a time traveler!

```suggestion
 * Copyright (C) 2022 Apple Inc. All Rights Reserved.
```

dcrousso committed:
Web Inspector: should not include `@globalPrivate` functions in stack traces
https://bugs.webkit.org/show_bug.cgi?id=242806

Reviewed by Yusuke Suzuki.

`@globalPrivate` are just internal implementation details, and therefore are likely to change (at
least more often than standardized functions) and be unknown to a developer.

* Source/JavaScriptCore/runtime/ImplementationVisibility.h: Added.
Create a new enum that (currently) allows indicating `Public` or `Private`.

* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_generator.py:
(BuiltinsGenerator.generate_embedded_code_data_for_function):
(BuiltinsGenerator.generate_embedded_code_string_section_for_data):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_generate_combined_header.py:
(BuiltinsCombinedHeaderGenerator.generate_forward_declarations):
(BuiltinsCombinedHeaderGenerator.generate_externs_for_object):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_generate_separate_header.py:
(BuiltinsSeparateHeaderGenerator.generate_externs_for_object):
* Source/JavaScriptCore/Scripts/wkbuiltins/builtins_templates.py:
Mark functions annotated with `@globalPrivate` as being `ImplementationVisibility::Private`.

* Source/JavaScriptCore/builtins/BuiltinUtils.h:
* Source/JavaScriptCore/builtins/BuiltinExecutableCreator.h:
* Source/JavaScriptCore/builtins/BuiltinExecutableCreator.cpp:
(JSC::createBuiltinExecutable):
* Source/JavaScriptCore/builtins/BuiltinExecutables.h:
* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::createDefaultConstructor):
(JSC::BuiltinExecutables::createBuiltinExecutable):
(JSC::BuiltinExecutables::createExecutable):
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.cpp:
(JSC::UnlinkedFunctionExecutable::UnlinkedFunctionExecutable):
Pass along the `ImplementationVisibility`.

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::makeFunction):
Assume `ImplementationVisibility::Public` for now.

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitNewClassFieldInitializerFunction):
Mark as `ImplementationVisibility::Private` as this is not supposed to be visible to developers.

* Source/JavaScriptCore/runtime/CodeCache.cpp:
(JSC::CodeCache::getUnlinkedGlobalFunctionExecutable):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(functionCreateBuiltin):
Mark as `ImplementationVisibility::Public`.

* Source/JavaScriptCore/runtime/ExecutableBase.h:
* Source/JavaScriptCore/runtime/ExecutableBaseInlines.h:
(JSC::ExecutableBase::implementationVisibility const): Added.
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
(JSC::FunctionExecutable::implementationVisibility const): Added.
Add a way to access the `ImplementationVisibility`.

* Source/JavaScriptCore/inspector/ScriptCallStackFactory.cpp:
(Inspector::CreateScriptCallStackFunctor::operator() const):
Ignore anything marked as `ImplementationVisibility::Private`.

* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.h:
* Source/JavaScriptCore/parser/Nodes.h:
* Source/JavaScriptCore/parser/ParserModes.h:
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
* Source/JavaScriptCore/runtime/ScriptExecutable.h:
Drive-by: Decrease the bit width of `LexicalScopeFeatures` to make room for `ImplementationVisibility`.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:

* LayoutTests/inspector/canvas/recording-bitmaprenderer-frameCount-expected.txt:
* LayoutTests/inspector/canvas/recording-bitmaprenderer-full-expected.txt:
* LayoutTests/inspector/canvas/recording-bitmaprenderer-memoryLimit-expected.txt:
* LayoutTests/inspector/console/message-stack-trace-expected.txt:
* LayoutTests/inspector/debugger/js-stacktrace-expected.txt:

Canonical link: https://commits.webkit.org/252554@main
